analyses/hotspots-detection/utils/filterMaf.R

```diff
@@ -0,0 +1,138 @@
+#' @return A dataframe with base columns from maf filtered as per user input for Variant_Classification
+#' column with "variant_classification" OR IMPACT column with "impact_values" OR Hugo_Symbol using "gene_table".
+#' Additional filtering can be done per amino acid position hotspots using "hotspot_amino_acid_site_df"
+#' OR per genomic site hotspots using "hotspot_genomic_site_df"
+#'
+
+filterMaf <- function(table,
+                      impact_values,
+                      hotspot_database_2017_snv_df,
+                      hotspot_database_2017_indel_df,
+                      hotspot_genomic_site_df) {
+  if (!missing(impact_values)) {
+    # IMPACT filtering values
+    table <- table %>%
+      dplyr::filter(IMPACT %in% impact_values)
+  }
+  
+  # filtered calls_base
+  calls_base <- table %>%
+    # only keep annotation for canonical protein_coding transcripts 
+    filter( CANONICAL=='YES' & 
+              BIOTYPE=='protein_coding') %>%
+    as.data.frame() %>%
+    # create an Amino_Acid_Position column
+    dplyr::mutate(
+      Amino_Acid_Position =
+        case_when(
+          # If not splice site we want to capture Amino_Acid_Position to match
+          !grepl("Splice_Site",Variant_Classification) ~
+            # format of Protein_position is 594/766
+            # [Amino_Acid_Position/ Protein_length]
+            # stripping protein length here to get Amino_Acid_Position
+            gsub("/.*", "", Protein_position),
+          # If splice site we want to capture HGVSp_Short to match 
+          grepl("Splice_Site",Variant_Classification) ~
+            # since these sites are intronic sites adding
+            # HGSVp_Short instead of Protein_position
+            gsub("p[.]","",HGVSp_Short))
+    ) %>%
+    mutate(
+      Start_Position = as.numeric(Start_Position),
+      End_Position = as.numeric(End_Position)
+    )
+
+  if (!missing(hotspot_database_2017_snv_df) & !missing(hotspot_database_2017_indel_df)) {
+    # filter by amino acid hotspot_database
+    calls_base_aa_filt_match <- calls_base %>%
+      dplyr::inner_join(hotspot_database_2017_snv_df,
+        by = c("Amino_Acid_Position", "Hugo_Symbol")
+      ) %>%
+      select(-Amino_Acid_Position)
+    
+    # if indel then check for overlap 
+    # within region of Amino_Acid_Position
+    calls_base_aa_filt_indel <- calls_base %>%
+      # if indel in Variant_Classification
+      filter(grepl("Ins|Del",Variant_Classification )) %>%
+      mutate( Amino_Acid_Start = gsub("-.*" , "", Amino_Acid_Position),
+              Amino_Acid_End = gsub ("*.-" , "" , Amino_Acid_Position)
+              ) %>%
+      dplyr:: left_join(hotspot_database_2017_indel_df,
+                        by = c( "Hugo_Symbol"),
+                        suffix = c(".calls_base",".hotspot_indel_df")  
+      ) %>% 
+      # keep only sites that where 
+      # calls$start >= hotspot_indel$start and
+      # calls$end <= hotspot_indel$stop
```

Comment:
In the V2 file they provide all of the calls that were used to define these ranges, and in no instance does INDEL extend beyond the start or stop. If the goal is to label hotspots INDELs as they do in the V2 file, the approach in this comment is appropriate.

Reply:
I am not sure I understand what you are saying here. We have both the hotspot file and a variety of MAF files that we are looking at. That is to say, we are comparing our indel calls to the ones in the hotspot, so we might well have some that extend outside the defined hotspots.

Either way, the logic I present will work for _any_ overlap (including fully contained ones), so unless we have reasons to exclude overlaps that are not fully enclosed by the defined hotspots, I think it would be prudent to switch.

Reply:
I'm talking about the original MAF that was used to construct these hotspot files: http://download.cbioportal.org/cancerhotspots/cancerhotspots.v2.maf.gz. The calls in that were then processed to create the INDEL and SNV inputs that Krutika is using to process what you have here. You can parse ranges for that if you like, but it's easier to see in the INDEL input. Using the example that Krutika gave above:

Hugo_Symbol | Amino_Acid_Position | Variant_Amino_Acid
-- | -- | --
CDKN2A | 27-42 | V28_E33del:4
CDKN2A | 27-42 | L32_L37del:3
CDKN2A | 27-42 | A36_N39delinsD:1
CDKN2A | 27-42 | E27del:1
CDKN2A | 27-42 | G35_A36del:1
CDKN2A | 27-42 | N39_N42del:1
CDKN2A | 27-42 | R29_A34del:1

Here you can see the range for one hotspot indel along with the contributing variants they recognized in their cohort MAF (formatted like `HGVSp:supporting variants count`). So here for the range starting at 27 and ending at 42, in no instance does the actual variant extend beyond the bounds of the range. This observation is true for ALL ranges presented in the INDEL input.

The reason I am providing for exclusion of variants that are not fully enclosed in the range is that data from MSKCC doesn't support INDELs that extend beyond the range. Is there a scientific reason to allow any overlap to count as a hotspot?

Reply:
> Is there a scientific reason to allow any overlap to count as a hotspot?

I would not necessarily count them "as a hotspot", but if the point of this analysis is to identify mutations in our dataset that might have been missed from earlier consensus, then yes, I would say that we should include anything that overlaps a hotspot. If we had a deletion in our data set from position 26-29 in the above example, I would think we would want to note it, even if it was not exactly one of the MSKCC identified mutations.

Reply:
Here's an example from the [V2 indel hotspot](https://github.com/AlexsLemonade/OpenPBTA-analysis/blob/81dbcd5b71161c225fb226cd86b71281e935dc1e/analyses/hotspots-detection/input/hotspots_database/hotspot_database_2017_indel.tsv) for CDKN2A, where 

Hugo_Symbol | Amino_Acid_Position | Variant_Amino_Acid
-- | -- | --
CDKN2A | 27-42 | V28_E33del:4
CDKN2A | 27-42 | L32_L37del:3
CDKN2A | 27-42 | A36_N39delinsD:1
CDKN2A | 27-42 | E27del:1
CDKN2A | 27-42 | G35_A36del:1
CDKN2A | 27-42 | N39_N42del:1
CDKN2A | 27-42 | R29_A34del:1

- `Amino_Acid_Position` is a region that fully overlaps each hotspot and none of the indels start of end outside the given region so we believe that the hotspot start and end AA positions  should be within the given region to be considered as a hotspot according to MSKCC database . 
- actual `Variant_Amino_Acid` found in samples, denoated as `<Variant change>:<number of samples that have the variant change >`

Reply:
Hi @jashapiro - we had initially thought about identifying novel hotspots, but then brought the scope of this back to only _known_ hotspots, so that is why we had thought to only include INDELS within the already defined hotspot region here. I agree that we would probably want to capture INDELS overlapping a part of this region, but perhaps that is beyond the scope because I would also want to verify those in IGV or molecularly in order to call them high-confidence. Does that sound ok to you? cc @jaclyn-taroni because I think Josh is out this week.

Reply:
> Hi @jashapiro - we had initially thought about identifying novel hotspots, but then brought the scope of this back to only _known_ hotspots, so that is why we had thought to only include INDELS within the already defined hotspot region here. I agree that we would probably want to capture INDELS overlapping a part of this region, but perhaps that is beyond the scope because I would also want to verify those in IGV or molecularly in order to call them high-confidence. Does that sound ok to you?

I understand the motivation not to add new indels that would require validation, but I don't think the current method accomplishes that, which leads me to think that we have two options:

- If we don't want to add any indels not already called, it seems we should be matching to the `Variant_Amino_Acid` call, and only keeping those precise locations. In my mind, that would be too stringent, as indels are notoriously difficult to call precisely. 

- If we are willing to accept calls that do not match the `Variant_Amino_Acid` locations, but do fall within `Amino_Acid_Position` ranges, then I think we have to also include overlaps that extend beyond that range. 
 
My argument is this: looking at [the table above](https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/956#discussion_r602476016), a deletion from position 30-38 would not match any of the variants in MSKCC, but would be caught by being within the 27-42 range. On the other hand, a deletion from 26-34 would fall outside the strict range criteria, but I can see no reason to exclude such a deletion if we were to see it.


I will note that there is a secondary question, which I did not think this PR was to address, which is how we determine which of the hotspot-overlapping calls are high confidence. I would doubt that a call only made by VarDict would be high confidence, even if it happened to appear in a known hotspot. As this PR looks only at callers in isolation, I was expecting that would be a later stage where the calls got merged.

Following on the above example, if there were a sample variant call made by one caller as 26-34del and by another caller as 27-34del, I would hope we would want to include it! If we exclude the former call by a strict overlap at this stage, then we might miss valid calls.

Reply:
Yes, these are good points, especially that we are not strictly matching now and that two callers may call the same INDEL with different coordinates. I am on board now :). How much of an overlap of this region do you think is then sufficient for calling an INDEL?

Reply:
> How much of an overlap of this region do you think is then sufficient for calling an INDEL?

This is a discussion point for the later PR. For this one, I think we can plan to make the suggested changes and re-request a review from @jashapiro.